File: PBProjects/Swift Maelstrom/maclib/XattrResFork.m
//
//  XattrResFork.m
//  Maelstrom
//
//  Created by C.W. Betts on 11/19/15.
//
//

#import "XattrResFork.h"
#include <sys/xattr.h>

typedef struct  {
	ssize_t pos;
	char *buffer;
	size_t size;
} fmem;

static int readfn(void * handler, char *buf, int size)
{
	fmem *mem = handler;
	size_t available = mem->size - mem->pos;
	
	if (size > available) {
		size = (int)available;
	}
	memcpy(buf, mem->buffer + mem->pos, sizeof(char) * size);
	mem->pos += size;
	
	return size;
}

static fpos_t seekfn(void *handler, fpos_t offset, int whence)
{
	size_t pos;
	fmem *mem = handler;
	
	switch (whence) {
		case SEEK_SET: {
			if (offset >= 0) {
				pos = (size_t)offset;
			} else {
				pos = 0;
			}
			break;
		}
		case SEEK_CUR: {
			if (offset >= 0 || (size_t)(-offset) <= mem->pos) {
				pos = mem->pos + (size_t)offset;
			} else {
				pos = 0;
			}
			break;
		}
		case SEEK_END: pos = mem->size + (size_t)offset; break;
		default: return -1;
	}
	
	if (pos > mem->size) {
		return -1;
	}
	
	mem->pos = pos;
	return (fpos_t)pos;
}

static int closefn(void *handler)
{
	fmem *f = handler;
	free(f->buffer);
	free(f);
	
	return 0;
}

FILE *fileFromResourceFork(NSURL *url)
{
	const char* fsr = url.absoluteURL.fileSystemRepresentation;
	ssize_t rsrcSize = getxattr(fsr, XATTR_RESOURCEFORK_NAME, NULL, 0, 0, 0);
	if (rsrcSize <= 0) {
		return NULL;
	}
	NSMutableData *mutData = [[NSMutableData alloc] initWithLength:rsrcSize];
	ssize_t gotBytes = getxattr(fsr, XATTR_RESOURCEFORK_NAME, mutData.mutableBytes, mutData.length, 0, 0);
	if (gotBytes == -1) {
		return NULL;
	}
	
	NSCAssert(rsrcSize == gotBytes, @"Different sizes?");
	
	char *bytes = malloc(mutData.length);
	[mutData getBytes:bytes length:mutData.length];
	
	fmem *aMem = malloc(sizeof(fmem));
	
	aMem->buffer = bytes;
	aMem->size = mutData.length;
	aMem->pos = 0;
	
	return funopen(aMem, readfn, NULL, seekfn, closefn);
}

File: PBProjects/Swift Maelstrom/maclib/XattrResFork.swift
//
//  XattrResFork.swift
//  Maelstrom
//
//  Created by C.W. Betts on 11/16/15.
//
//

import Foundation
import Darwin.POSIX.sys.xattr

private struct Fmem {
	var pos: Int = 0
	var buffer: Unmanaged<NSData>
	
	var size: Int {
		return buffer.takeUnretainedValue().length
	}
}

private func readfn(handler: UnsafeMutablePointer<Void>, buf: UnsafeMutablePointer<Int8>, var size: Int32) -> Int32 {
	let mem = UnsafeMutablePointer<Fmem>(handler)
	let available = mem.memory.size - mem.memory.pos;
	
	if Int(size) > available {
		size = Int32(available)
	}
	memcpy(buf, mem.memory.buffer.takeUnretainedValue().bytes.advancedBy(mem.memory.pos), sizeof(Int8) * Int(size));
	mem.memory.pos += Int(size);
	
	return size;
}

private func seekfn(handler: UnsafeMutablePointer<Void>, offset: fpos_t, whence: Int32) -> fpos_t {
	let mem = UnsafeMutablePointer<Fmem>(handler)
	var pos = 0
	
	switch (whence) {
	case SEEK_SET:
		if offset >= 0 {
			pos = Int(offset)
		} else {
			pos = 0;
		}
		
	case SEEK_CUR:
		if (offset >= 0 || Int(-offset) <= mem.memory.pos) {
			pos = mem.memory.pos + Int(offset);
		} else {
			pos = 0;
		}
		
	case SEEK_END:
		pos = mem.memory.size + Int(offset)
	default:
		return -1;
	}
	
	if (pos > mem.memory.size) {
		return -1;
	}
	
	mem.memory.pos = pos;
	return fpos_t(pos)
}

private func closefn(handler: UnsafeMutablePointer<Void>) -> Int32 {
	let mem = UnsafeMutablePointer<Fmem>(handler)
	
	mem.memory.buffer.release()
	mem.dealloc(1)
	
	return 0
}

func fileFromResourceFork(url: NSURL) -> UnsafeMutablePointer<FILE> {
	#if os(OSX)
		let fsr = url.absoluteURL.fileSystemRepresentation
		let rsrcSize = getxattr(fsr, XATTR_RESOURCEFORK_NAME, nil, 0, 0, 0)
		guard rsrcSize > 0 else {
			return nil
		}
		let mutData = NSMutableData(length: rsrcSize)!
		let gotBytes = getxattr(fsr, XATTR_RESOURCEFORK_NAME, mutData.mutableBytes, mutData.length, 0, 0)
		guard gotBytes != -1 else {
			return nil
		}
		
		assert(rsrcSize == gotBytes)
		
		let fmemPtr = UnsafeMutablePointer<Fmem>.alloc(1)
		fmemPtr.memory.pos = 0
		let storedBytes = NSData(data: mutData)
		fmemPtr.memory.buffer = Unmanaged.passRetained(storedBytes)
		
		return funopen(fmemPtr, readfn, nil, seekfn, closefn)
	#else
		return nil
	#endif
}
